### DISCUSSION TITLE
How to ensure packer.startup runs before require("package")?

### DISCUSSION THREAD

**Discussion by steve-lorimer** [wbthomason/packer.nvim#651] [Q&A] [answered]
I am installing a package (`nvim-mapper`) using packer:

**`lua/plugins.lua`**:

```lua
packer.startup({
    function(use)

      ...

      use({
          "lazytanuki/nvim-mapper",
          config = function()
            require("nvim-mapper").setup({})
          end,
          module = "nvim-mapper",
      })

      ...

    end
})
```

In another file, I want to `require("nvim-mapper")` and call some functions therein:

**`lua/keybindings.lua`**:

```lua
local mapper = require("nvim-mapper")

local map = function(mode, lhs, rhs, opts, category, unique_identifier, description)
  mapper.map(mode, lhs, rhs, opts, category, unique_identifier, description)
end
```

My main `init.lua` file just requires the above 2 files:

**`init.lua`**:

```lua
...

require("plugins")
require("keybindings")
```

However, even though I've `required` the plugins file **before** the keybindings file, neovim complains on startup:

```
Error detected while processing /home/steve/.config/nvim/init.lua:
E5113: Error while calling lua chunk: /home/steve/.config/nvim/lua/keybindings.lua:1: module 'nvim-mapper' not found:
        no field package.preload['nvim-mapper']
        no file './nvim-mapper.lua'
        no file '/home/runner/work/neovim/neovim/.deps/usr/share/luajit-2.1.0-beta3/nvim-mapper.lua'
        no file '/usr/local/share/lua/5.1/nvim-mapper.lua'
        no file '/usr/local/share/lua/5.1/nvim-mapper/init.lua'
        no file '/home/runner/work/neovim/neovim/.deps/usr/share/lua/5.1/nvim-mapper.lua'
        no file '/home/runner/work/neovim/neovim/.deps/usr/share/lua/5.1/nvim-mapper/init.lua'
        no file './nvim-mapper.so'
        no file '/usr/local/lib/lua/5.1/nvim-mapper.so'
        no file '/home/runner/work/neovim/neovim/.deps/usr/lib/lua/5.1/nvim-mapper.so'
        no file '/usr/local/lib/lua/5.1/loadall.so'
```

How can I ensure that the `use(... nvim-mapper)` call has completed before my keybindings code runs and attempts to load it?

**Comment by wbthomason** [comment, marked as answer]
This is due to the order in which (Neo)vim sources files at startup.

`packer`, by default, outputs its loader code to `plugin/packer_compiled.lua`. This code includes the setup necessary for the `module` lazy-loader to work. The files in `plugin/` are all automatically sourced during startup, but *after* your `init.lua` runs - so, your `require` fails.

There's a long-standing plan to move sourcing of the lazy-loaders to the very start of the startup process, but, until I have time to implement that large change, you have two options for workarounds:
1. Put your `require('keybindings')` in a file that runs after `plugin/packer_compiled.lua`, e.g. `plugin/x_keybindings.lua` or similar (ordering is lexicographic).
2. Use the `config.compile_path` configuration variable to change where `packer_compiled.lua` is output to a directory which is **not** automatically sourced by Neovim, and manually source it yourself at the top of your init file, e.g. `config.compile_path = ~/.config/nvim/lua/packer_compiled.lua` and `require('packer_compiled')` in your init file.

I hope this helps!

**Comment by steve-lorimer** [comment]
Thanks for the clarification!

Prior to your above answer, I found the section in your docs about the `PackerComplete` event, and created an `autocmd` to call a function to continue loading other modules once the plugins are loaded.

Below is the code:

```
require 'plugins'

_G.plugins_loaded = function()
  require 'keybindings'
end

vim.cmd [[autocmd User PackerComplete ++once lua plugins_loaded()]]
```

It seems to work, although is perhaps not optimal... not sure? Any comments on my so called "solution"?

Thanks again!

**Reply by wbthomason** [reply]
Huh, I'm actually kind of surprised that this approach works - `PackerComplete` should only fire after operations like `update`, `sync`, etc. - not on the conclusion of `startup`. So, really, I'm surprised that your `plugins_loaded` function ever runs.

**Reply by steve-lorimer** [reply]
Ah, that's good to know, thanks!

So it turns out it's working for me because I also automatically run `PackerSync` on startup... 

```
    if vim.tbl_count(vim.fn.globpath(vim.fn.stdpath("data") .. "/site/pack/packer/opt", "*", 0, 1)) == 1 then
      vim.cmd("PackerSync")
    else
      vim.cmd("PackerClean")
      vim.cmd("PackerInstall")
    end
```

When my config is stable I would have removed the this and everything would have stoped working! :facepalm:
